### PR TITLE
Add auto-incrementing of version  for Experiment

### PR DESCRIPTION
**Pull request by DatCorno**

### CODE REVIEW

File: src/orion/core/evc/conflicts.py [hunk 1]
```diff
@@ -1553,44 +1553,70 @@ def __init__(self, conflict, new_name):
                 The conflict which is resolved by this resolution.
             new_name: string
                 A new name for the branching experiment. A ValueError is raised if name is already
-                in database.
+                in database with a direct child.
```
Comment (bouthilx) [inline src/orion/core/evc/conflicts.py]: new_version is missing in the doc here.

File: src/orion/core/evc/conflicts.py [hunk 2]
```diff
@@ -1553,44 +1553,70 @@ def __init__(self, conflict, new_name):
...
+                # Since the name changes, we reset the version count.
+                self.new_version = 1
+
+            # If the new name is the same as the old name, we are trying to increment
+            # the version of the experiment.
+            elif self._check_for_children():
                 raise ValueError(
                     "Experiment name \'{0}\' already exist for user \'{1}\'".format(
-                        new_name, self.conflict.username))
+                        self.new_name, self.conflict.username))
```
Comment (bouthilx) [inline src/orion/core/evc/conflicts.py]: `and has children. Version cannot be auto-incremented and a new name is required for branching.

File: src/orion/core/evc/conflicts.py [hunk 3]
```diff
@@ -1553,44 +1553,70 @@ def __init__(self, conflict, new_name):
...
-            query = {'name': name, 'metadata.user': self.conflict.username}
+            query = {'name': self.new_name, 'metadata.user': self.conflict.username}
 
             named_experiments = Database().count('experiments', query)
             return named_experiments == 0
 
+        def _check_for_children(self):
+            """Check if experiment has children"""
+            # If we made it this far, new_name is actually the name of the parent.
+            query = {'name': self.new_name, 'metadata.user': self.conflict.username}
```
Comment (bouthilx) [inline src/orion/core/evc/conflicts.py]: This query can be dropped

File: tests/unittests/core/evc/test_conflicts.py [hunk 4]
```diff
@@ -372,6 +360,52 @@ def test_try_resolve(self, experiment_name_conflict):
...
+
+    def test_conflict_exp_no_child(self, exp_no_child_conflict):
+        """Verify the version number is incremented when exp has no child."""
+        new_name = "test"
+        assert not exp_no_child_conflict.is_resolved
+        resolution = exp_no_child_conflict.try_resolve(new_name)
+        assert isinstance(resolution, exp_no_child_conflict.ExperimentNameResolution)
+        assert exp_no_child_conflict.is_resolved
+        assert resolution.conflict is exp_no_child_conflict
+        assert resolution.new_version == 2
```
Comment (bouthilx) [inline tests/unittests/core/evc/test_conflicts.py]: should we test that `exp_no_child_conflict.old_config['version'] == 1` to make sure there is indeed an increment?

File: tests/unittests/core/evc/test_conflicts.py [hunk 5]
```diff
@@ -372,6 +360,52 @@ def test_try_resolve(self, experiment_name_conflict):
...
+        new_name = "test"
+        assert not exp_w_child_conflict.is_resolved
+        resolution = exp_w_child_conflict.try_resolve(new_name)
+        assert isinstance(resolution, exp_w_child_conflict.ExperimentNameResolution)
+        assert exp_w_child_conflict.is_resolved
+        assert resolution.conflict is exp_w_child_conflict
+        assert resolution.new_version == 3
+
+    def test_conflict_exp_w_child_as_parent(self, exp_w_child_as_parent_conflict):
+        """Verify the version number is incremented from child when exp has a child."""
```
Comment (bouthilx) [inline tests/unittests/core/evc/test_conflicts.py]: Docstring here is from test above

File: tests/unittests/core/evc/test_conflicts.py [hunk 6]
```diff
@@ -372,6 +360,52 @@ def test_try_resolve(self, experiment_name_conflict):
...
+            exp_w_child_as_parent_conflict.try_resolve(new_name)
+
+        assert "Experiment name" in str(exc.value)
+
+    def test_conflict_exp_renamed(self, exp_w_child_as_parent_conflict):
+        """Verify the version number is not incremented when exp is renamed."""
+        # It increments from child
+        new_name = "test2"
+        assert not exp_w_child_as_parent_conflict.is_resolved
+        resolution = exp_w_child_as_parent_conflict.try_resolve(new_name)
```
Comment (bouthilx) [inline tests/unittests/core/evc/test_conflicts.py]: Same here for testing version number before test. Would be good to use an old_config where version > 1 to make sure version is actually set to 1 and not just kept as is.

File: tests/unittests/core/io/interactive_commands/test_branching_prompt.py [hunk 7]
```diff
@@ -458,6 +459,7 @@ def test_reset_exp_name(self, conflicts, branch_solver_prompt):
         branch_solver_prompt.do_reset("'{}'".format(str(conflicts.get_resolved()[0].resolution)))
         assert len(conflicts.get_resolved()) == 0
 
+    @pytest.mark.skip(reason="Fix once --version is introduced")
```
Comment (bouthilx) [inline tests/unittests/core/io/interactive_commands/test_branching_prompt.py]: We would this need --version?

File: tests/unittests/core/test_branch_config.py [hunk 8]
```diff
@@ -407,22 +408,12 @@ def test_reset_dimension(self, parent_config, new_config):
         assert isinstance(conflict, NewDimensionConflict)
         assert len(conflicts.get_resolved()) == 0
 
+    @pytest.mark.skip(reason="Reinstate once --version has been added.")
     def test_name_experiment(self, parent_config, child_config, create_db_instance):
-        """Test if changing the experiment names work for valid name"""
+        """Test if having the same experiment name does not create a conflict."""
```
Comment (bouthilx) [inline tests/unittests/core/test_branch_config.py]: This should be turned into a case where we branch from parent which has a child, and then the test stays the same, changing the name we resolve the conflict.

File: tests/unittests/core/test_branch_config.py [hunk 9]
```diff
@@ -438,11 +429,11 @@ def test_bad_name_experiment(self, parent_config, child_config, create_db_instan
 
         conflict = conflicts.get([ExperimentNameConflict])[0]
 
-        conflict.new_config['name'] = 'should-not-be-overwritten'
+        conflict.old_config['name'] = 'should-not-be-overwritten'
         assert not conflict.is_resolved
         branch_builder.change_experiment_name('test')
         assert len(conflicts.get_resolved()) == 0
-        assert conflict.new_config['name'] == 'should-not-be-overwritten'
+        assert conflict.old_config['name'] == 'should-not-be-overwritten'
```
Comment (bouthilx) [inline tests/unittests/core/test_branch_config.py]: This should be new_config. It's to test that the resolution was reverted properly.

File: tests/unittests/core/test_branch_config.py [hunk 10]
```diff
@@ -735,6 +726,7 @@ def test_rename_missing_changed_marked(self, parent_config, child_config):
         assert resolved_conflicts[0].resolution.conflict.dimension.name == '/x'
         assert resolved_conflicts[0].resolution.new_dimension_conflict.dimension.name == '/w_b'
 
+    @pytest.mark.skip(reason="Reinstate this test once --version is added.")
     def test_name_experiment(self, parent_config, child_config, create_db_instance):
```
Comment (bouthilx) [inline tests/unittests/core/test_branch_config.py]: Same comment here for the skip

File: tests/unittests/core/test_branch_config.py [hunk 11]
```diff
@@ -438,7 +441,7 @@ def test_bad_name_experiment(self, parent_config, child_config, create_db_instan
 
         conflict = conflicts.get([ExperimentNameConflict])[0]
 
-        conflict.new_config['name'] = 'should-not-be-overwritten'
+        conflict.old_config['name'] = 'should-not-be-overwritten'
```
Comment (bouthilx) [inline tests/unittests/core/test_branch_config.py]: This one should be old_config. :thinking:

File: tests/unittests/core/io/interactive_commands/test_branching_prompt.py (hunk 7 above)
Comment (bouthilx) [reply tests/unittests/core/io/interactive_commands/test_branching_prompt.py]: Skip still needed?

**Comment by DatCorno** [comment]
[Make exp version tag auto-increment](https://trello.com/c/T89Bgs3M/176-make-exp-version-tag-auto-increment)

File: tests/unittests/core/test_branch_config.py [hunk 12]
```diff
@@ -751,17 +762,6 @@ def test_name_experiment(self, parent_config, child_config, create_db_instance):
         assert conflict.new_config['name'] == new_name
         assert conflict.is_resolved
 
-    def test_bad_name_experiment(self, parent_config, child_config, create_db_instance):
-        """Test if experiment name conflict is not resolved when invalid name is marked"""
```
Comment (bouthilx) [inline tests/unittests/core/test_branch_config.py]: Why remove this one?

**Pull request merged**